File: lib/golden_record/public_id.ex
defmodule GoldenRecord.PublicId do
  alias GoldenRecord.{Events, Substrate, Survivorship, Api}

  @moduledoc """
  Identity-grade, customer-facing schemes like CNK. The surrogate key is internal; CNK is the
  public key — strictly unique, never shared. Two sources giving different CNKs for the same
  product is fine: they become canonical + alias(es) on one key, with the canonical chosen by
  priority. The customer resolves by ANY of them.
  """

  @doc "Canonical public id of `scheme` for `key` (by source priority), plus its aliases."
  def canonical(scheme, key, log, priority),
    do: do_canonical(scheme, key, Api.members(log), fn -> identity_claims(log) end, priority)

  @doc """
  canonical/4 with the fold state prebuilt — `members` (`Api.members/1`) and the current
  identity claims — so callers enriching many keys fold the log once instead of per key.
  """
  def canonical(scheme, key, members, idclaims, priority),
    do: do_canonical(scheme, key, members, fn -> idclaims end, priority)

  # idclaims arrives as a thunk: the log arity only pays for the identity-claims fold on the
  # multi-code path, exactly as before.
  defp do_canonical(scheme, key, members, idclaims_fun, priority) do
    codes =
      members
      |> Map.get(key, MapSet.new())
      |> Enum.filter(fn {s, _} -> s == scheme end)
      |> Enum.sort()

    case codes do
      [] ->
        nil

      [code] ->
        %{canonical: code, aliases: []}

      _multiple ->
        idclaims = idclaims_fun.()

        entries =
          for code <- codes, src <- sources_of(code, idclaims), do: %{source: src, value: code, order: 0}

        case entries do
          [] ->
            %{
              canonical: nil,
              aliases: codes,
              status: :needs_review,
              candidates: Enum.map(codes, &{nil, &1})
            }

          _ ->
            decision = Survivorship.decide(scheme, entries, priority)

            if decision.status == :needs_review do
              %{
                canonical: nil,
                aliases: codes,
                status: :needs_review,
                candidates: decision.candidates
              }
            else
              %{canonical: decision.value, aliases: List.delete(codes, decision.value)}
            end
        end
    end
  end

  @doc "Identity-grade INVARIANT check: a code of `scheme` must never own >1 key. Returns violations."
  def collisions(scheme, log) when is_list(log), do: collisions(scheme, Api.members(log))

  def collisions(scheme, members) when is_map(members) do
    members
    |> Enum.flat_map(fn {k, codes} -> for {s, _} = c <- codes, s == scheme, do: {c, k} end)
    |> Enum.group_by(fn {c, _} -> c end, fn {_, k} -> k end)
    |> Enum.filter(fn {_c, keys} -> length(Enum.uniq(keys)) > 1 end)
    |> Enum.map(fn {c, keys} -> %{code: c, keys: Enum.sort(Enum.uniq(keys))} end)
  end

  defp sources_of(code, idclaims), do: for(c <- idclaims, code in c.data.codes, do: c.source)

  @doc "Current identity claims — the second piece of fold state canonical/5 takes prebuilt."
  def identity_claims(log),
    do: for(%Events.ClaimAsserted{kind: :identity} = e <- log, do: e) |> Substrate.current()
end

File: php/src/PublicId.php
<?php

declare(strict_types=1);

namespace Ingot;

/**
 * Identity-grade public ids (CNK) — ported from `PublicId` in lib/golden_record_core.ex.
 *
 * The surrogate key is internal; CNK is the public key — strictly unique, never shared. Two
 * sources giving different CNKs for one product become canonical + alias(es) on one key (canonical
 * chosen by source priority). `collisions` is the hard invariant: a code of a scheme must never own
 * more than one key.
 */
final class PublicId
{
    /**
     * Canonical public id of `scheme` for `key` (by source priority), plus its aliases, or null.
     *
     * @param list<DomainEvent> $log
     * @return array<string,mixed>|null
     */
    public static function canonical(string $scheme, string $key, array $log, Priority|callable $priority): ?array
    {
        $members = self::ledger($log)->members[$key] ?? [];
        $codes = [];
        foreach (Sets::values($members) as $code) {
            if ($code[0] === $scheme) {
                $codes[] = $code;
            }
        }
        usort($codes, static fn (array $a, array $b): int => strcmp(Codes::key($a), Codes::key($b)));

        if ($codes === []) {
            return null;
        }

        if (count($codes) === 1) {
            return ['canonical' => $codes[0], 'aliases' => []];
        }

        $idclaims = self::identityClaims($log);

        $entries = [];
        foreach ($codes as $code) {
            foreach (self::sourcesOf($code, $idclaims) as $src) {
                $entries[] = ['source' => $src, 'value' => $code, 'order' => 0];
            }
        }

        if ($entries === []) {
            return [
                'canonical' => null,
                'aliases' => $codes,
                'status' => 'needs_review',
                'candidates' => array_map(static fn (array $code): array => [null, $code], $codes),
            ];
        }

        $decision = Survivorship::decide($scheme, $entries, $priority);
        if ($decision->needsReview()) {
            return [
                'canonical' => null,
                'aliases' => $codes,
                'status' => 'needs_review',
                'candidates' => $decision->candidates,
            ];
        }

        return [
            'canonical' => $decision->value,
            'aliases' => array_values(array_filter($codes, static fn (array $code): bool => $code !== $decision->value)),
        ];
    }

    /**
     * Identity-grade invariant check: a code of `scheme` must never own >1 key. Returns violations.
     *
     * @param list<DomainEvent> $log
     * @return list<array{code: array{0: string, 1: string}, keys: list<string>}>
     */
    public static function collisions(string $scheme, array $log): array
    {
        $members = self::ledger($log)->members;

        /** @var array<string, array{code: array{0: string, 1: string}, keys: array<string,true>}> $byCode */
        $byCode = [];
        $codeOrder = [];
        foreach ($members as $k => $codes) {
            foreach (Sets::values($codes) as $c) {
                if ($c[0] !== $scheme) {
                    continue;
                }
                $ck = Codes::key($c);
                if (!isset($byCode[$ck])) {
                    $byCode[$ck] = ['code' => $c, 'keys' => []];
                    $codeOrder[] = $ck;
                }
                $byCode[$ck]['keys'][$k] = true;
            }
        }

        $out = [];
        foreach ($codeOrder as $ck) {
            $keys = array_keys($byCode[$ck]['keys']);
            if (count($keys) > 1) {
                sort($keys, SORT_STRING);
                $out[] = ['code' => $byCode[$ck]['code'], 'keys' => $keys];
            }
        }

        return $out;
    }

    /**
     * @param array{0: string, 1: string} $code
     * @param list<Claim> $idclaims
     * @return list<string>
     */
    private static function sourcesOf(array $code, array $idclaims): array
    {
        $out = [];
        foreach ($idclaims as $c) {
            foreach ($c->data['codes'] as $cc) {
                if ($cc === $code) {
                    $out[] = $c->source;
                    break;
                }
            }
        }

        return $out;
    }

    /**
     * @param list<DomainEvent> $log
     * @return list<Claim>
     */
    private static function identityClaims(array $log): array
    {
        $claims = [];
        foreach ($log as $e) {
            if ($e instanceof Claim && $e->kind === 'identity') {
                $claims[] = $e;
            }
        }

        return Substrate::current($claims);
    }

    /**
     * @param list<DomainEvent> $log
     */
    private static function ledger(array $log): LedgerState
    {
        $state = IdentityLedger::new();
        foreach ($log as $e) {
            $state = IdentityLedger::evolve($state, $e);
        }

        return $state;
    }
}
